doc/lexer-rules.md

```diff
@@ -308,3 +309,21 @@ As of 4.5, you can also define channel names like enumerations with the followin
 ```
 channels { WSCHANNEL, MYHIDDEN }
 ```
+
+### caseInsensitive()
```

Comment:
hmm... We should probably discuss change it such as this before you spend time on it. I'm not sure we need to add this part. Maybe we just fix the bug first and then worry about other changes?  What is the motivation? I think you were arguing against this previously

Reply:
> hmm... We should probably discuss change it such as this before you spend time on it.

No problem, it was interesting for me how it's simple to implement such a feature.

> Maybe we just fix the bug first and then worry about other changes?

It seems I've fixed all bugs by the previous PR (at least I've tested all case insensitive grammars).

> What is the motivation? I think you were arguing against this previously

I've encountered a couple of rules where `caseInsensitive` should be disabled in case insensitive grammars during testing all grammars-v4. The first one is described in related issue: #3436. The second one looks like [the following](https://github.com/antlr/grammars-v4/blob/master/sql/postgresql/PostgreSQLLexer.g4#L2577):

```g4
fragment EscapeStringText
    : 'u' letterOrDigit letterOrDigit letterOrDigit letterOrDigit
    | 'U' letterOrDigit letterOrDigit letterOrDigit letterOrDigit letterOrDigit letterOrDigit letterOrDigit letterOrDigit
    ;
fragment letterOrDigit: [0-9a-fA-F];
```

Also, there are some [Unicode ranges](https://github.com/antlr/grammars-v4/blob/master/sql/tsql/TSqlLexer.g4#L941-L956) that ideally should not be transformed (but they can be rewritten a bit to get rid of `CHARACTERS_COLLISION_IN_SET` warning):

```g4
fragment FullWidthLetter
    : '\u00c0'..'\u00d6'
    | '\u00d8'..'\u00f6'
    | '\u00f8'..'\u00ff'
    | '\u0100'..'\u1fff'
    | '\u2c00'..'\u2fff'
    | '\u3040'..'\u318f'
    | '\u3300'..'\u337f'
    | '\u3400'..'\u3fff'
    | '\u4e00'..'\u9fff'
    | '\ua000'..'\ud7ff'
    | '\uf900'..'\ufaff'
    | '\uff00'..'\ufff0'
    ;
```

Reply:
I wonder if a case insensitive option on the rule isn't better because this really isn't an action.

Reply:
You are right, it's isn't an action. On the other side, it looks quite natural and clear in my option. Also, it's the simplest implementation without grammar changing (at least I don't know how it could be implemented better).

Reply:
well I would disagree that it looks natural because only commands go after `->` :)    If we add an option I don't think we need to change the grammar right?

Reply:
meaning rules already have options, if I remember correctly.

Reply:
Indeed, it seems like we have rule-level options "the tool still supports the following syntax for future use": https://github.com/antlr/antlr4/blob/master/doc/options.md#rule-options

Thanks, I'll try to reimplement the rule-level `caseInsensitive` option this way.

Reply:
Unfortunately, it looks like only parser rules support such options but not lexer rules, `parserRule` contains `rulePrequels` that contains `optionsSpec`, but `lexerRule` does not contain `rulePrequels`. That's why grammar should be changed a bit.

Reply:
oh dang! Hmm...let me look. crap. That would take major work (tree grammar too). Not sure that's worth the risk. I'm not in favor of making an action though either. Maybe we hold off on this one and think about later. Might need lexer rule options at some point.

Reply:
I'll try to implement it using lexer rules. I've already estimated it seems not so major work. Anyway, I'll suggest a solution and you'll make the decision.

Reply:
I've remade it using the lexer rule option, and now it looks really more conscious than my first implementation with command, thanks @parrt! And it was not so complicated,